latch/registry/utils.py

```diff
@@ -0,0 +1,343 @@
+import json
+import keyword
+import re
+from enum import Enum
+from typing import Annotated, Dict, List, Optional, TypeVar, Union
+
+from latch.gql.execute import execute
+from latch.types import LatchDir, LatchFile
+
+T = TypeVar("T")
+
+clean_regex = re.compile("\W|^(?=\d)", re.UNICODE)
+
+account_root_regex = re.compile("^account_root\/(?P<root>[^/]+)(?P<path>\/.*)$")
+mount_regex = re.compile("^mount\/(?P<mount>[^/]+)(?P<path>\/.*)$")
+shared_regex = re.compile("^shared(?P<path>\/.*)$")
+
+
+def clean(name: str):
+    # https://stackoverflow.com/a/3305731
+    cleaned = clean_regex.sub("_", name)
+
+    if keyword.iskeyword(cleaned):
+        cleaned = f"_{cleaned}"
+    return cleaned
+
+
+def to_python_type(
+    registry_type: Dict[str, Dict],
+    allow_empty: bool = False,
+):
+    ret = None
+    if "primitive" in registry_type:
+        primitive = registry_type["primitive"]
+        if primitive in ["string", "datetime", "date"]:
+            ret = str
+        elif primitive == "integer":
+            ret = int
+        elif primitive == "number":
+            ret = float
+        elif primitive == "blob":
+            if "metadata" in registry_type:
+                if "nodeType" in registry_type["metadata"]:
+                    if registry_type["metadata"]["nodeType"] == "dir":
+                        ret = LatchDir
+            ret = LatchFile
+        elif primitive == "link":
+            from latch.registry.record import Record
+
+            ret = Record
+        elif primitive == "enum":
+            ret = Enum("Enum", registry_type["members"])
+        elif primitive == "null":
+            ret = type(None)
+        elif primitive == "boolean":
+            ret = bool
+        else:
+            raise ValueError(f"invalid primitive type: {registry_type['primitive']}")
+
+    elif "array" in registry_type:
+        ret = List[to_python_type(registry_type["array"], False)]
+
+    elif "union" in registry_type:
+        variants = []
+        for key, variant in registry_type["union"].items():
+            variants.append(
+                Annotated[
+                    to_python_type(
+                        variant,
+                        False,
+                    ),
+                    key,
```

Comment:
not sure what you mean, `typing.Annotated` takes in variadic args so u can have as many annotations as u want per type (cf. https://docs.python.org/3/library/typing.html#typing.Annotated)

also why do we need more than one annotation here? seems enough to just annotate the type with its corresponding tag

Reply:
ok nvm it's just flattened

still we want to carry a tag (by wrapping in a dataclass) since there is no way of telling what the annotation is other than by type